Golem rebalancing & code refactor; also, some new signals

Metekillot opened the pull request:
## About The Pull Request

See #88750
This change makes golems more useful, a little more deadly, but also a bit more vulnerable.

### Plasma & Plasteel no longer give pressure resistance, pressure resistance is innate to golems
### Plasteel now gives golems TRAIT_NOCRITDAMAGE, which is a... trait that... means you don't take damage in crit...
### Plastitanium now gives what Plasteel does, along with TRAIT_NOSOFTCRIT and TRAIT_NOHARDCRIT. You'll stay standing until you drop dead. However, _**you will still suffer slowdown from health loss** and **you cannot heal with iron while using a mineral effect**_.
### Golems now have RESISTCOLD as well. But a caveat. Pressure that is too low will cause your joints to begin coldwelding together due to uh, oxidization or uh, something... anyway, it'll make you real slow, so don't start busting windows because you won't die to space or you're easy pickings.
https://github.com/user-attachments/assets/8de1c53b-ea39-4635-bbdf-3607d526a43d


# Uranium now causes your punches to be radioactive (and your hugs, sometimes (20% of the time to be exact), but you didn't mean to!). Irradiating someone with this will lower the duration of your uranium effect each time by 30 seconds.

Also I added 4 new signals if you even fucking care
```
/// Called from /datum/species/proc/harm() (unarmed attack chain is the Bad Place)
/// Sent whenever you perform a harmful attack with a bodypart
/// SEND_SIGNAL(target, signal, user, target, user.get_active_hand, damage_received, attacked_bodypart)
#define COMSIG_CARBON_BODYPART_ATTACK_PERFORMED "bodypart_attack_performed"
/// As above, but sent to the victim of such unprovoked violence
/// SEND_SIGNAL(target, signal, user, target, user.get_active_hand, damage_received, attacked_bodypart)
#define COMSIG_CARBON_BODYPART_ATTACK_VICTIM "bodypart_attack_victim"
/// Ideally would be called anywhere unarmed damage is performed but it will be an entire project
/// to sort out the unarmed attack chain
/// Note the inclusion of how much damage is received

/// Called from /mob/living/carbon/human/help_shake_act (help_shake_act click chain is the Bad Place)
/// Also ideally called anywhere you are touching someone, barehanded, without lethal intent
/// SEND_SIGNAL(user, signal, user, target, user.get_active_hand)
#define COMSIG_CARBON_BODYPART_CONTACT_PERFORMED "bodypart_contact_performed"
/// As above, but when you are the one to be touched tenderly
/// SEND_SIGNAL(target, signal, user, target, user.get_active_hand)
#define COMSIG_CARBON_BODYPART_CONTACT_RECEIVED "bodypart_contact_received"
```
## Why It's Good For The Game

I don't see very many golems and I, myself, don't often play golems, because micromanaging the mineral effects is a pain in the ass and most of them... kinda suck. Uranium most of all; its effect is literally "you don't have to deal with this shit for 5 minutes" which doesn't seem like inspired game-design. So I made golems higher-utility for things like rescuing crewmembers from breaches or exploring space ruins, without them becoming windowbreaking mass murderers who vent the whole station before Sec finishes their bedhead_2 lesbian roleplaying.

The uranium thing is also very funny, extremely telegraphed, and you look like Shrek whenever you have the uranium effect active, so if you're running around with it active the immense suspicion and very likely kill-on-sight you'll suffer seem to be a fair balance for chernobyl shenanigans; lessening the duration by a significant amount each time you (would) successfully cause the effect also seems quite fair, so you don't eat one sheet of uranium and embark on a game of mass green-tag.

You also still aren't immune to heat damage (unless you eat plasma); this has no effect on wading through lava because who cares, go sit in the lava.

I also added the signals and cleaned up the golem status effect code because it was a mess.
## Changelog
:cl: Bisar Metek
add: Uranium infused golems now irradiate with their punches (or their hugs, sometimes).
balance: Plasteel and plasma no longer grant low and high pressure immunity to golems, respectively.
balance: Golems are now immune to pressure damage and to cold damage, innately.
add: Golems in hazardously low pressure now experience cold-welding in their joints, causing them to become extremely slow.
refactor: The code for golem status effects and arm effects has been cleaned up.
/:cl:

Metekillot committed:
shuffling some code around so that one status effect doesn't have a snowflake proc on it, changing some of the status effects, changing some golem innate traits

Metekillot committed:
new effects for golems, new penalties for golems, new traits for golems... and signals

Holinka4ever commented:
I am kinda happy that atleast golems do not die in space! Woo hoo!

Metekillot committed:
finishing touches

Learner10 commented:
Interesting. If merged, I'll try to make plenty of golems to acquire some more samples and input on the balance. Plus... uranium golems accidentally causing issue seems like an on-point trend for SS13.

timothymtorres commented:
These sound amazing and give golems more soul. +1

LSommers commented:
NuGolems getting some much-needed love.

Metekillot committed:
Merge branch 'master' into golems-for-the-ungrateful

Jacquerel commented on code/__DEFINES/dcs/signals/signals_mob/signals_mob_carbon.dm:
```diff
@@ -173,3 +173,22 @@
+/// As above, but sent to the victim of such unprovoked violence
+/// SEND_SIGNAL(target, signal, user, target, user.get_active_hand, damage_received, attacked_bodypart)
+#define COMSIG_CARBON_BODYPART_ATTACK_VICTIM "bodypart_attack_victim"
+/// Ideally would be called anywhere unarmed damage is performed but it will be an entire project
+/// to sort out the unarmed attack chain
+/// Note the inclusion of how much damage is received
```
Doc comments will be applied to whatever the following line is, if this relates to the above define then move them up.

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -132,12 +159,49 @@
+		)
+		ASYNC // async so we can scream without upsetting should_not_sleep
+			target.emote("scream")
+	remove_duration(30 SECONDS) // lessen the duration to avoid uranium golems becoming walking chernobyl
```
Might be worth warning the golem in some way that this is happening. 
Text in the chat is sufficient, but pulsing the border of the status effect (or some other visual) too would be cool.

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -95,6 +108,20 @@
+	for(var/obj/item/bodypart/arm/arm in human_owner.bodyparts)
+		if(arm.limb_id != SPECIES_GOLEM)
+			continue
+		LAZYADD(modified_arms, arm)
```
Should probably also have the memory management qdel signal handling from the diamond ore (perhaps alos others?) here instead.  
If the arm is somehow deleted it will still otherwise exist in this list and cause a hard deletion, unless you track it being deleted and remove it from the list at this point.

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -362,28 +432,26 @@
+		arm.unarmed_damage_low -= damage_increase
+		arm.unarmed_damage_high -= damage_increase
+		UnregisterSignal(arm, COMSIG_QDELETING)
+	. = ..()
```
Some people prefer `return ..()` if it's the end of a proc anyway, though I'm not enforcing this as a requirement.

Jacquerel commented on code/modules/mob/living/carbon/human/_species.dm:
```diff
@@ -1043,7 +1045,7 @@ GLOBAL_LIST_EMPTY(features_by_species)
 		return // dont attack after
 	if(owner.combat_mode)
-		harm(owner, target, attacker_style)
+		harm(owner, target, attacker_style) // would like to put signals for attacking out here but the harm proc is a disaster
```
Probably not a useful or necessary comment

Jacquerel reviewed (changes requested):
I'm a little wary of the permanent pressure resistance but players really really hate slowdown so that might be a sufficient alternative.  
Possibly consider removing that debuff when they are consuming plasteel (or whatever the old spacewalking mineral was) because they cannot wear spacesuits so there's actually no other way around this which _would_ allow them to go off into space. Plus the thing you replaced it with isn't all that exciting by itself.  
  
I would personally be tempted to make cold-welding build up over time but that's maybe just because I overengineer everything.

Metekillot committed:
changing the comments

Metekillot committed:
added some pretty little flashy flashes and also made them able to refresh their mineral duration

Metekillot commented:
@Jacquerel I also changed the golem status effect to being a refresh instead of a unique type, so they can main a mineral type if they want, instead of having to carry around a backpack of minerals to snack on every few minutes to refresh a given effect. I figure locking them into the duration is a fair trade-off to not have to deal with the micromanagement.

Jacquerel commented:
> @Jacquerel I also changed the golem status effect to being a refresh instead of a unique type, so they can main a mineral type if they want, instead of having to carry around a backpack of minerals to snack on every few minutes to refresh a given effect. I figure locking them into the duration (and no iron healing) is a fair trade-off to not have to deal with the micromanagement.

Oranges was pretty firm in the original design document that it should not work this way and this was what afaik he last said on the matter
https://github.com/tgstation/tgstation/pull/74197#issuecomment-1496691110
As he specifically spoke about not wanting this specific change I wouldn't want to merge it until you've talked to him about it, which is one of the unfortunate reasons not to decide to add new features to a PR mid-review 😉

Metekillot commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -95,13 +93,33 @@
 
-// Ensure that a human_owner is set
+/datum/status_effect/golem/tick(seconds_between_ticks)
+	var/static/warning = FALSE
```
No, you're right, I got tunnel vision and was trying some harebrained pre-optimization and forgetting how status effects worked.

Metekillot commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -200,6 +227,22 @@
 			target.emote("scream")
 	remove_duration(30 SECONDS) // lessen the duration to avoid uranium golems becoming walking chernobyl
+	var/target_name_word = splittext("[target]", " ")[1] // retrieve the first word of their displayed name
+	owner.balloon_alert_to_viewers("irradiated [target_name_word]!")
```
I didn't read anything in the hackmd for feedback nor in the balloon alert comments about the '!' bit but having a ! here isn't a make or break. Regarding showing this to all viewers: yes, it's intentional that there's a balloon alert to notify everyone nearby, because irradiation is a huge effect

Jacquerel replied:
preferably we would replace this with a short visual effect rather than relying on balloon alert to viewer then, could we do like a beam that lasts one second or something?
balloon alerts describing what just happened should really be kind of an absolute last resort

Metekillot replied:
fair enough, i'd like to dig into methods for safely wordifying stuff in the future but that's a pin to put in the future

Metekillot commented:
> > @Jacquerel I also changed the golem status effect to being a refresh instead of a unique type, so they can main a mineral type if they want, instead of having to carry around a backpack of minerals to snack on every few minutes to refresh a given effect. I figure locking them into the duration (and no iron healing) is a fair trade-off to not have to deal with the micromanagement.
> 
> Oranges was pretty firm in the original design document that it should not work this way and this was what afaik he last said on the matter [#74197 (comment)](https://github.com/tgstation/tgstation/pull/74197#issuecomment-1496691110) As he specifically spoke about not wanting this specific change I wouldn't want to merge it until you've talked to him about it, which is one of the unfortunate reasons not to decide to add new features to a PR mid-review 😉
> 
> I'll be honest though I don't see how it refreshing on consumption prevents you from needing to carry the minerals around, if it's just resetting the duration then you would need to eat the same minerals at the same interval in order to achieve the same coverage of time, it just gives more of a grace period where you do not spend a small amount of time with the effect inactive.

I read orange's comments and I saw him expressing the same sentiment I intended ( get locked in if you choose such and such mineral ) which was what my refresh change was intended to allow people to double down on, but I don't feel like splitting hairs on it if agonizing micromanagement is the design doc vision so I trimmed it back out. Rest of the review changes are in, too! Cheers

Metekillot committed:
changes from review

Metekillot committed:
removed vars and replaced with constants

Metekillot committed:
satisfying CI

Jacquerel commented:
> Possibly consider removing that debuff when they are consuming plasteel (or whatever the old spacewalking mineral was) because they cannot wear spacesuits so there's actually no other way around this which _would_ allow them to go off into space. Plus the thing you replaced it with isn't all that exciting by itself.

Any reason you don't want to do this? I can add it myself, but as you haven't mentioned it I don't know if you had a specific reason other than "don't want to"

Metekillot committed:
made a visual effect for golems applying the uranium effect

Metekillot commented:
> > Possibly consider removing that debuff when they are consuming plasteel (or whatever the old spacewalking mineral was) because they cannot wear spacesuits so there's actually no other way around this which _would_ allow them to go off into space. Plus the thing you replaced it with isn't all that exciting by itself.
> 
> Any reason you don't want to do this? I can add it myself, but as you haven't mentioned it I don't know if you had a specific reason other than "don't want to"

i will get to it soon, i didn't expect getting a simple visual effect to take me 4 hours of wrestling with how the renderer works.

Metekillot committed:
plasteel golems are less slowed by cold welding; animated effects communicate this

Metekillot commented:
Added some visual effects to the plasteel status alert and the accompanying pressure alert for golem coldwelding. Added visual effects for uranium punching as well.

Metekillot committed:
state name 'none' -> 'nothing'

Metekillot committed:
i forgot to update the name after updating the name...

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_food_buff.dm:
```diff
@@ -62,7 +62,11 @@
 /datum/golem_food_buff/plasteel
 	status_effect = /datum/status_effect/golem/plasteel
-	added_info = "If consumed this mineral will harden you against the hazards of space."
+	added_info = "If consumed this mineral will harden your body, keeping you from slipping into death from critical health."
```
Should probably return to mentioning the space thing if you re-added that.

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -73,28 +78,79 @@
 	return TRUE
 
+/datum/status_effect/golem/proc/species_handling(datum/source, datum/species/new_species, datum/species/old_species)
+	SIGNAL_HANDLER
+	SHOULD_CALL_PARENT(TRUE)
```
Why is this on the base datum when only one effect uses it, and why must it call parent when parent does nothing?

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -73,28 +78,79 @@
+	var/atom/movable/screen/alert/status_effect/golem_status/status_alert = linked_alert
+	if(!status_alert)
+		src.tick_interval = STATUS_EFFECT_NO_TICK
+		log_admin("[src] status effect has no alert, stopping tick.")
```
Is this a debug log you forgot to remove? I don't think admins would care about this.

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -132,12 +188,68 @@
+		vision_distance = COMBAT_MESSAGE_RANGE
+		)
+	remove_duration(30 SECONDS) // lessen the duration to avoid uranium golems becoming walking chernobyl
+	ASYNC // async so we can scream without upsetting should_not_sleep
+		target.emote("scream")
```
```suggestion
	INVOKE_ASYNC(target, TYPE_PROC_REF(/mob, emote), "scream")
```
I think this adds some safety checks.

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -213,25 +323,141 @@
+	applied_fluff = "Plasteel plates reinforce your body and your will!"
 	alert_icon_state = "sheet-plasteel"
-	alert_desc = "You are sealed against the cold, and against low pressure environments."
+	alert_desc = "You won't take damage while in crit, and you're resistant (but not immune) to the cold-welding effect of low pressure. Healing out of crit is another problem..."
```
Same here wrt slowness resistance.

Jacquerel commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -213,25 +323,141 @@
+	stop_pressure_slow_resistance_mitigation()
+	return ..()
+
+/datum/status_effect/golem/plasteel/proc/start_pressure_slow_resistance_mitigation()
```
This implementation is a bit messy, to make a suggestion as to how to reorganise it:

- Keep most of this code in `/datum/movespeed_modifier/status_effect/golem_coldwelding`
- Inside the coldwelding debuff, register signals for addition or removal of a trait applied by this status effect and use that to decide how slow the golem should be (if at all). Unfortunately `TRAIT_STATUS_EFFECT(id)` probably won't work because all golem status effects have the same id.
- Also check for that trait on initial application.
- You probably then don't need to care about handling species changes because you can also just do that in the cold welding effect and have it qdel itself.

Jacquerel commented on code/modules/mob/living/carbon/human/species_types/golems.dm:
```diff
@@ -100,3 +103,40 @@
+	desc = "The lack of atmosphere is preventing your joints from oxidizing apart; they're welding together and it's harder to move!"
+	icon_state = "golem_statued"
+
+/datum/species/golem/handle_environment_pressure(mob/living/carbon/human/golem, datum/gas_mixture/environment, seconds_per_tick, times_fired)
```
Oh this isn't a status effect already, oops.
Well in that case add onto my previous description:

- Make "coldwelding is active" into a status effect (conveniently will also handle its own screen alert for you).
- This proc only cares about applying the status effect if the golem is insufficiently pressurised.
- The status effect monitors if the golem enters an area of normal pressure, and deletes itself.
- It also monitors whether you gain/lose the plasteel effect and adjusts the slowdown appropriately

That should make all of this logic a bit simpler I think?
Also technically probably I should be asking for this to be on a golem's brain or torso or something because we prefer not to have this kind of stuff on the species datum but tbh any of those other places would be equally arbitrary.

Jacquerel reviewed (changes requested):
Sorry for the delay I didn't notice that you'd added the thing I asked for, oops.
